### DISCUSSION TITLE
using dataclasses as node and edge attributes

### DISCUSSION THREAD

**Discussion by FGFischer** [networkx/networkx#5607] [Q&A]
I couldnt find anything about using dataclasses combined with networkx, which is why i dont know if this is a good practice.

My question is: Is this a good practice? Is NetworkX meant to be used like this? 

Since networkx already has a "dict-of-dicts"-like datastructure i assume that my approach has some disadvantages since i add another "level of depth" to the datastructure (dict-of-dicts-of-dataclasses).
I assume functions like "[to_pandas_Dataframe](https://networkx.org/documentation/networkx-1.10/reference/generated/networkx.convert_matrix.to_pandas_dataframe.html)" dont work. the dataclass would have to be iteratble.
 On the other hand it works as intended (i think), i can use specific classmethods and it is very clear which data is used by what function.

A little example of what did:
- lets assume i have an already written [dataclass](https://docs.python.org/3/library/dataclasses.html), which works fine as it is, it has lots of useful functionality and i dont want to change it.

```
#Example Dataclass
from dataclasses import dataclass, field
@dataclass()
class data_cluster():
    data1:str
    data2:int
    data3:np.ndarray
    data4:dict = field(init=False)
    def __post_init__(self):
        self.data4 = {self.data1:self.data2}
    
    def func1(self,G,u_of_edge):
        #do something based on weights of the edge/in_edge/out_edge/u_node/v_node
        self.data1 = G.nodes[u_of_edge]['node_weight']
        self.data4[self.data1] = self.data2

```

- Then i initilize e.g. a directed Graph as seen below with a dataclass as weight, instead of multiple weights, for the edge.

```

#Example Network 
import networkx as nx
G = nx.DiGraph()

#add dataclass as weight to network
G.add_edge(u_of_edge='A',v_of_edge='B', weight = data_cluster(data1='foo', data2='1.0', data3 = np.zeros(5)))
G.add_node(node_for_adding='A',node_weight="bar")
```
and do something like this.
```
#Example Function Dataclass <> DiGraph
for node in nx.topological_sort(G): #['A','B'] "list"
    for edge in G.out_edges(node): # ('A','B') "tuple"
         print(G.edges[edge]['weight'])
         G.edges[edge]['weight'].func1(G=G,u_of_edge=edge[0]) 
         print(G.edges[edge]['weight']) 
```
Output:
```
data_cluster(data1='bar', data2='1.0', data3=array([0., 0., 0., 0., 0.]), data4={'foo': '1.0', 'bar': '1.0'})
data_cluster(data1='foo', data2='1.0', data3=array([0., 0., 0., 0., 0.]), data4={'foo': '1.0'})
```

**Comment by MridulS** [comment]
On the surface of it, I think it should work just fine. The dataclass defined is pretty much a dictionary of data (in my head atleast) while we interact with it. Adding a `__getitem__` to your dataclass should also make accessing attributes via `[ ]` possible.

Even `pandas` read write should work just fine, it doesn't need the attribute itself to be an iterable. 

If anything does break do let us know in this thread :)

PS: If you are using `to_pandas_dataframe` please consider upgrading your networkx version. It was deprecated and removed a long time back :)

**Comment by dschult** [comment]
We have a number of functions to move node and edge values between the graphs and other data structures such as `pandas.DataFrame`, `numpy.ndarray` and `scipy.sparse.sparse_array`.  There is also a PR #4217 with some nice functions to convert node and edge data to/from other data structures in a unified general way. 

But often it is better to write your own data storage and access methods to suit your data and your needs. The approach you describe seems pretty good.
